sklearn/neural_network/tests/test_mlp.py

```diff
@@ -662,6 +662,23 @@ def test_warm_start():
             clf.fit(X, y_i)
 
 
+@pytest.mark.parametrize("MLPEstimator", [MLPClassifier, MLPRegressor])
+def test_warm_start_full_iteration(MLPEstimator):
+    # Non-regression test for:
+    # https://github.com/scikit-learn/scikit-learn/issues/16812
+    # Check that the MLP estimator accomplish `max_iter` with a
+    # warm started estimator.
+    X, y = X_iris, y_iris
+    max_iter = 3
+    clf = MLPEstimator(
+        hidden_layer_sizes=2, solver='sgd', warm_start=True, max_iter=max_iter
+    )
+    clf.fit(X, y)
+    assert max_iter == clf.n_iter_
+    clf.fit(X, y)
+    assert max_iter == clf.n_iter_
```

Comment:
The use case of warm_start is to allow for the increase of `max_iter` to add more iterations:

```py
clf.fit(X, y)
assert clf.n_iter_ == 3

# keeps the first 3 iterations and trains for 4 more iterations.
clf.set_params(max_iter=7)
clf.fit(X, y)
assert clf.n_iter_ == 7
```

In the case of `warm_start=True` and an unchanged `max_iter`, I would suspect the second call to `fit` would be a noop.

Reply:
Thanks for your comment. I'm (again) away from kbd in the mountains, but AFAIR this is not the case. If warm start is true the starting classes are initialized with the outcome of the last fit. That would mean that the second call to fit here with three iterations will still give different results. 

But please correct me if I'm wrong, not having the code at hand now.

Reply:
@thomasjpfan thanks for your comment. I see how this makes sense in random forests, but I don't see how - or what - we could change/improve here at the moment.

We can change the number of max iters, but that will just give another different model, since in total the model went not through 3+3, but through 3+n iterations. While this can be made a test, I don't see any difference in the **check** we can make, since there is nothing similar to the `n_estimators`.

To spell it differently: in my restricted knowledge, it seems that warm starting without changing the number of estimators will not change the outcome in the case of random forests, but as far as I see it will change the outcome int he case of mlpclassifiers - or am I wrong here? That said, I am not really sure what kind of improvement you would like to see here, besides giving different number of max iterations for the second call?

Reply:
`n_estimators` and `max_iter` are the parameters that are warm started for their respective estimators. But the implementation of warm_start in MLP is different. It can warm start without changing the parameters, which needs to be discussed outside of this PR.

That being said, the `n_iter_` behavior for `MLPRegressor` is correct:

### master

```py
from sklearn.neural_network import MLPRegressor
from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split
X, y = make_regression(n_samples=200, random_state=1)
regr = MLPRegressor(max_iter=10, warm_start=True).fit(X, y)

assert regr.n_iter_ == 10

regr.fit(X, y)
assert regr.n_iter_ == 20
```

### This PR

```py
from sklearn.neural_network import MLPRegressor
from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split
X, y = make_regression(n_samples=200, random_state=1)
regr = MLPRegressor(max_iter=10, warm_start=True).fit(X, y)

assert regr.n_iter_ == 10

regr.fit(X, y)
assert regr.n_iter_ == 10
```

From the original issue: https://github.com/scikit-learn/scikit-learn/issues/16812 we want to make sure that `MLPClassifier` also follows the behavior of `MLPRegressor` + `n_iter_`.

> That said, I am not really sure what kind of improvement you would like to see here, besides giving different number of max iterations for the second call?

Yes, preserving the behavior on master regarding `n_iter_` is what I am suggesting.

Reply:
Thanks for your patience, now I understand the issue, and I am aware of it.

That was in fact a conscious change by myself, because I consider the original behavior (shown above in master branch) to increase the number of iterations done when `incremental == False` as incorrect.

I assumed - probably wrongly - that `incremental == False` should mean we start training from the beginning (`n_iter_ = 0`) for the full set of max iterations. 

Orthogonal to this, if `warm_start == True`, then the starting model is not random, but the outcome from a previous fit.

But these two are orthogonal, meaning, that if `warm_start == True` and `incremental == False` we start at iteration 0, etc.

Furthermore, the only distinguishable difference is not in the actual outcome, but only in the value of `n_iter_`, other than this there are no behavioral differences.

I guess the bottomline is the following question of mine: What is the expected/correct behavior of `n_iter` under `incremental == False` when calling `fit`?

Reply:
To be more specific: My initial proposal contained this code
```
--- a/sklearn/neural_network/_multilayer_perceptron.py
+++ b/sklearn/neural_network/_multilayer_perceptron.py
@@ -379,6 +379,11 @@ class BaseMultilayerPerceptron(BaseEstimator, metaclass=ABCMeta):
                                            incremental):
             # First time training the model
             self._initialize(y, layer_units, X.dtype)
+        # In case we are warm starting but aren't incremental, the
+        # above routine is not called and the n_iter_ counter is not reset,
+        # do it here.
+        if not incremental:
+            self.n_iter_ = 0
```

related to my above statements.

Reverting the current state of this PR to the behavior of master is easy with these changes:
```
diff --git a/sklearn/neural_network/_multilayer_perceptron.py b/sklearn/neural_network/_multilayer_perceptron.py
index bdab6b49f..2937e59e3 100644
--- a/sklearn/neural_network/_multilayer_perceptron.py
+++ b/sklearn/neural_network/_multilayer_perceptron.py
@@ -379,10 +379,6 @@ class BaseMultilayerPerceptron(BaseEstimator, metaclass=ABCMeta):
                                            incremental):
             # First time training the model
             self._initialize(y, layer_units, X.dtype)
-        if self.warm_start and not incremental:
-            # only the number of iteration should be initialized since
-            # weights are already allocated due to the warm start.
-            self.n_iter_ = 0
 
         # Initialize lists
         activations = [X] + [None] * (len(layer_units) - 1)
diff --git a/sklearn/neural_network/tests/test_mlp.py b/sklearn/neural_network/tests/test_mlp.py
index f8ff68445..72f347966 100644
--- a/sklearn/neural_network/tests/test_mlp.py
+++ b/sklearn/neural_network/tests/test_mlp.py
@@ -676,7 +676,7 @@ def test_warm_start_full_iteration(MLPEstimator):
     clf.fit(X, y)
     assert max_iter == clf.n_iter_
     clf.fit(X, y)
-    assert max_iter == clf.n_iter_
+    assert 2 * max_iter == clf.n_iter_
 
 
 def test_n_iter_no_change():
```

I am happily pushing this to the PR, please let me know.

Reply:
I think the change of behavior of `n_iter_` shouldn't be indeed part of the PR, and if we want to change this behavior, it should be done in a separate PR.

Thus, I have pushed the change and reverted to the behavior as in master with respect to `n_iter_`.

Thanks for you detailed considerations!